### PR TITLE
Support optional inputs in model signatures

### PR DESCRIPTION
**Pull request by apurva-koti**
## Related Issues/PRs

 #xxx

## What changes are proposed in this pull request?

Add an `optional` boolean parameter to `ColSpec` to specify whether the column in question is required for model inference or can be omitted. Updates code accordingly:
- Updates pyfunc schema checks to check against missing _required_ columns. Missing optional columns will be ignored, but provided optional columns will still be type checked.
- Defaults `optional` to `False` for backwards compatibility.
- Prevents argument autofill in pyfunc models returned as spark UDFs as added here https://github.com/mlflow/mlflow/pull/4236

Example usage (pyfunc):
```python
input_schema = Schema(
    [
        ColSpec("double", "a"),
        ColSpec("double", "b"),
        ColSpec("string", "c", optional=True),
        ColSpec("long", "d", optional=True),
    ]
)
signature = ModelSignature(inputs=input_schema)
data = {"a": [1.0], "b": [1.0]}
data_2 = {"a": [1.0], "b": [1.0], "d": [2]}

for data in [
    data,
    data_2,
]:
    pd_data = pd.DataFrame(data)
    check = _enforce_schema(pd_data, signature.inputs) #passes
```

Example (spark udf):
```python
test_signature = {
    "inputs": '[{"name": "a", "type": "long"}, {"name": "b", "type": "long"}, {"name" : "c", "type": "long", "optional": "True"}]',
}
signature = ModelSignature.from_dict(test_signature)
...
udf = mlflow.pyfunc.spark_udf(...)

data = spark.createDataFrame(
    pd.DataFrame(columns=["a", "b"], data={"a": [1, 2], "b": [2, 3]})
)

res = data.withColumn("response", udf(*data.columns)) #
```

## How is this patch tested?

- [x] Existing unit/integration tests
- [ ] New unit/integration tests
- [x] Manual tests (describe details, including test results, below)

## Does this PR change the documentation?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Make sure the changed pages / sections render correctly in the documentation preview.

## Release Notes
Optional input columns can now be specified in model signatures. These columns can be omitted from input dataframes at prediction time.

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

Optional input columns can now be specified in model signatures. These columns can be omitted from input dataframes at prediction time.

### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [x] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

**Comment by WeichenXu123** [comment]
Could you update PR description to attach an example code that shows the case your PR supports ?

**Comment by apurva-koti** [comment]
@WeichenXu123 PR not ready yet. I'll have all that in when requesting review

File: mlflow/pyfunc/__init__.py [hunk 1]
```diff
@@ -1301,7 +1301,12 @@ def batch_predict_fn(pdf):
     def udf_with_default_cols(*args):
         if len(args) == 0:
             input_schema = model_metadata.get_input_schema()
-
+            if input_schema and len(input_schema.optional_input_names()) > 0:
+                raise MlflowException(
+                    message="Cannot apply UDF without column names specified when"
+                    " model signature contains optional columns.",
+                    error_code=INVALID_PARAMETER_VALUE,
+                )
```
[the remark this answers is not in the document]
Comment (WeichenXu123) [reply mlflow/pyfunc/__init__.py:1310]: Make sense!

**Review by WeichenXu123 — APPROVED** [review]
LGTM

**Pull request merged**